modules/caddyhttp/fileserver/matcher.go

```diff
@@ -183,55 +180,91 @@ func (m MatchFile) selectFile(r *http.Request) (rel, abs string, matched bool) {
 		m.TryFiles = []string{r.URL.Path}
 	}
 
+	// common preparation of the file into parts
+	prepareFilePath := func(file string) (string, string) {
+		suffix := m.firstSplit(path.Clean(repl.ReplaceAll(file, "")))
+		if strings.HasSuffix(file, "/") {
+			suffix += "/"
+		}
+		fullpath := sanitizedPathJoin(root, suffix)
+		return suffix, fullpath
+	}
+
+	// sets up the placeholders for the matched file
+	setPlaceholders := func(info os.FileInfo, rel string, abs string) {
+		repl.Set("http.matchers.file.relative", rel)
+		repl.Set("http.matchers.file.absolute", abs)
+
+		fileType := "file"
+		if info.IsDir() {
+			fileType = "directory"
+		}
+		repl.Set("http.matchers.file.type", fileType)
+	}
+
 	switch m.TryPolicy {
 	case "", tryPolicyFirstExist:
 		for _, f := range m.TryFiles {
-			suffix := m.firstSplit(path.Clean(repl.ReplaceAll(f, "")))
-			fullpath := sanitizedPathJoin(root, suffix)
-			if strictFileExists(fullpath) {
-				return suffix, fullpath, true
+			suffix, fullpath := prepareFilePath(f)
+			info, err := os.Stat(fullpath)
+			if err != nil {
+				// in reality, this can be any error
+				// such as permission or even obscure
+				// ones like "is not a directory" (when
+				// trying to stat a file within a file);
+				// in those cases we can't be sure if
+				// the file exists, so we just treat any
+				// error as if it does not exist; see
+				// https://stackoverflow.com/a/12518877/1048862
+				continue
+			}
```

Comment:
Why was this brought out of `strictFileExists`?

Reply:
Because I need the `fileInfo` for `setPlaceholders`. I guess I could just return the `info` and the boolean together because golang allows for that :man_shrugging: I guess that's my PHP habits showing?